fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -1589,6 +1656,28 @@ ThreadFuture<Void> MultiVersionDatabase::waitPurgeGranulesComplete(const KeyRef&
 	return abortableFuture(f, dbState->dbVar->get().onChange);
 }
 
+ThreadFuture<bool> MultiVersionDatabase::blobbifyRange(const KeyRangeRef& keyRange) {
+	auto f = dbState->db ? dbState->db->blobbifyRange(keyRange) : ThreadFuture<bool>(Never());
+	return abortableFuture(f, dbState->dbVar->get().onChange);
+}
+
+ThreadFuture<bool> MultiVersionDatabase::unblobbifyRange(const KeyRangeRef& keyRange) {
+	auto f = dbState->db ? dbState->db->unblobbifyRange(keyRange) : ThreadFuture<bool>(Never());
+	return abortableFuture(f, dbState->dbVar->get().onChange);
+}
+
+ThreadFuture<Standalone<VectorRef<KeyRangeRef>>> MultiVersionDatabase::listBlobbifiedRanges(const KeyRangeRef& keyRange,
+                                                                                            int rowLimit) {
+	auto f = dbState->db ? dbState->db->listBlobbifiedRanges(keyRange, rowLimit)
+	                     : ThreadFuture<Standalone<VectorRef<KeyRangeRef>>>(Never());
+	return abortableFuture(f, dbState->dbVar->get().onChange);
+}
+
+ThreadFuture<Version> MultiVersionDatabase::verifyBlobRange(const KeyRangeRef& keyRange, Optional<Version> version) {
+	auto f = dbState->db ? dbState->db->verifyBlobRange(keyRange, version) : ThreadFuture<Version>(Never());
+	return abortableFuture(f, dbState->dbVar->get().onChange);
```

Comment:
Do any of these new database operations have a way to set a timeout on them?

Reply:
It doesn't seem like we have a good way of doing timeouts on database operations in general, as none of the existing database functions have them either.

Reply:
Any asynchronous operation should probably have some way to set timeouts. Initially I had planned to create some database operations for tenants, and they would have involved using the default timeout set on a database. We could also consider having a separate option that controls timeouts on DB operations if we needed more control.

In the end I didn't end up using this logic because I didn't add database level operations, but you can see the initial version of my idea here: https://github.com/apple/foundationdb/pull/6284/files#diff-83ebc47ff74810b627faf4cf5ef4fe089cfa3e5c98c9d4b7d4d8f69715dc2af9

Basically, there is some helper function that would let you run a DB-level operation and would automatically retry version changed errors and handle timeouts.

Reply:
All of the other database level operations that are asynchronous are somewhat new and don't really seem to be fully baked in the MVC, so I wouldn't rely too much on copying their implementations.

Reply:
At the very least, this should be changed to do something like the following:

```
auto dbVar = dbState->dbVar->get();
auto f = dbVar.value ? dbVar.value->verifyBlobRange(keyRange, version) : ThreadFuture<Version>(Never());
return abortableFuture(f, dbVar.onChange);
```

Otherwise, the `onChange` variable you are using in the last line may not correspond to the same database that you are calling your function on.

With that change, I think your function will at least be callable and return an error every time the version changes (which, by the way, can happen when you first connect to a cluster). It will not retry this error automatically, breaking the intended seamlessness of the error, though I don't know if semantically your operation would support a straightforward retry or not. It also sounds like it won't support timeouts, which we should (at least eventually) have for all asynchronous operations exposed through the C API.

Reply:
Good catch with the race, we should fix that.
I agree that adding timeouts seems good for the MVC Database calls in general, but given how large this PR already is/the scope of that expanding past these functions, I think doing that in a later PR would make more sense

Reply:
I'm ok to merge these in an incomplete state, and if we're going to do that I would recommend adding a comment (and  maybe even a follow-up GitHub issue) to track the completion of the API.

Reply:
Done.